purchase_update_proposal/models/purchase_order.py

```diff
@@ -0,0 +1,212 @@
+        """Customize according to your needs according delegation
+        attributed to users with only this group group_purchase_proposal_user
+        """
+        if initial_state in ("approved", "confirmed"):
+            self.signal_workflow("purchase_confirm")
+            self.signal_workflow("purchase_approved")
+
+    def _prepare_proposal_data(self):
+        self.ensure_one()
+        res = defaultdict(list)
+        for elm in self.proposal_ids:
+            if not elm.qty and not elm.price_u and not elm.date:
+                raise UserError(
+                    _("You must fill at least Qty, Date or Price on proposal")
+                )
+            vals = {"product_qty": elm.qty}
+            if elm.line_id in res:
+                # we already have a purchase_line as origin of these data
+                # then we'll create a new line
+                vals["order_id"] = elm.order_id.id
+                vals["price_unit"] = elm.price_u or elm.line_id.price_unit
+                vals["date_planned"] = elm.date or elm.line_id.date_planned
+                vals["product_id"] = elm.line_id.product_id.id
+                vals["name"] = elm.line_id.product_id.display_name
+                vals["proposal"] = elm
+            else:
+                # it'll be used for write
+                if elm.price_u:
+                    vals["price_unit"] = elm.price_u
+                if elm.date:
+                    vals["date_planned"] = elm.date
+                # We check original data are not the same than new ones
+                updated_vals = {x: vals[x] for x in vals if vals[x] != elm[x]}
+                vals = updated_vals
+            res[elm.line_id].append(vals)
+        self._check_data2update(res)
+        return res
+
+    def _check_data2update(self, data):
+        for key, val in data.items():
+            if isinstance(val, list) and not val[0]:
+                raise UserError(
+                    _("No data to update for line ID '%s'") % key.id)
+
+    def _update_proposal_to_purchase_line(self, data, body):
+        for line_id in data:
+            # we update first line
+            line_id.write(data[line_id][0])
+            body.append(_("Updated line '%s' with %s") % (line_id.id, data[line_id][0]))
+            if len(data[line_id]) > 1:
+                todo = len(data[line_id]) - 1
+                while todo:
+                    # create other lines if any
+                    self.env["purchase.order.line"].create(data[line_id][todo])
+                    body.append(_("Created line: %s" % data[line_id][todo]))
+                    todo -= 1
+
+    @api.multi
+    def write(self, vals):
+        if not self._get_purchase_groups() and [
+                x for x in vals.keys() if x[:9] != "proposal_"]:
+            # The user is not an Odoo purchaser, we must prevent to update other fields
+            logger.info("Fields being writing %s" % vals.keys())
+            raise UserError(
```

Comment:
I am not sure about that, seems a bit strong...
I mean, I guess there are several possibilites than a none purchaser may edit some fields on purchase...

Maybe with a custom group, maybe using sudo (ok superuser should probably be a purchaser), maybe using the suspend security module (in that case I think it will still raise here)...

Maybe it is possible to check the write access on `purchase.order` of the current user and raise if write access is denied instead the purchase groups ?

Reply:
I made it overridable but not sure it's complete your expectations !?

Reply:
Not really, I think we can do better so there is no need to override this depending on other modules.
Don't you think that check write rights make more sense ?
Something like :` if not self.check_access_rights('write', raise_exception=False)` instead of `if not self._get_purchase_groups()`

Reply:
In fact I have to raise if I have rights for write but not for some fields.
If I have no rights for write it'll raise later in the code. I don't know what you really mean ?